Basket creation endpoint

rlucioni opened the pull request:
This PR splits the ecommerce API across v1 and v2 packages in preparation for the deprecation of v1. The new basket creation endpoint is part of the v2 package. The payment processor list endpoint and the endpoint used for retrieval of an order by basket ID are also part of the v2 package.

The payment processors are left in a less-than-ideal state with the understanding that they are being overhauled in #58.

@clintonb and @jimabramson, could you please review this? @wedaly, FYI.

clintonb commented on ecommerce/extensions/api/v1/app.py:
```diff
@@ -1,6 +1,6 @@
 from oscar.core.application import Application
 
-from ecommerce.extensions.api.urls import urlpatterns
```
This file isn't necessary. This isn't an Oscar app, so we should be able to reference `api.urls` as we would in `urls.py` for any standard Django project.

rlucioni replied:
I've removed this file. Thanks for catching this.

clintonb commented on ecommerce/extensions/api/v2/authentication.py:
```diff
@@ -0,0 +1,2 @@
+"""JWT authentication scheme for use with DRF."""  # pylint: disable=unused-import
+from ecommerce.extensions.api.v1.authentication import JwtAuthentication, BearerAuthentication
```
Shared components such as `authentication`, `constants`, and `throttling` should live in the `api` package.

clintonb commented on ecommerce/extensions/urls.py:
```diff
@@ -16,7 +17,8 @@
 
     # Oscar URLs
     url(r'^i18n/', include('django.conf.urls.i18n')),
-    url(r'^api/v1/', include(api.urls)),
```
Keep this line, and declare your version paths in `api.urls`. Here is an example: http://james.lin.net.nz/2014/02/18/django-rest-framework-api-versioning/.

clintonb commented on ecommerce/settings/_oscar.py:
```diff
@@ -17,7 +17,8 @@
 
 # APP CONFIGURATION
 OSCAR_APPS = [
-    'ecommerce.extensions.api',
```
You should only need the root `api` package. The settings should not care about versions since we use URL versioning.

rlucioni replied:
You're right, I've corrected this as well.

clintonb commented on ecommerce/settings/_oscar.py:
```diff
@@ -125,7 +121,7 @@
         'access_key': 'set-me-please',
         'secret_key': 'set-me-please',
-        'pay_endpoint': 'https://replace-me/',
+        'payment_page_url': 'https://replace-me/',
```
What's the difference?

rlucioni replied:
`payment_page_url` is a more clear, explicit name that matches the other two keys corresponding to URLs, `receipt_page_url` and `cancel_page_url`.

clintonb commented:
@rlucioni You versioned a bit too much of the code. As I mentioned in another comment, you don't need to version `authentication`, `throttling`, etc. You really only need versions of `views`, `urls`, and the associated tests. Let me know if you have questions.

rlucioni commented:
@clintonb thank you for the feedback. I've addressed your comments by moving shared API modules into the root `api` package and using namespaces more effectively to version the URLs (e.g., the path returned by `reverse('api:v2:baskets:create')` maps to the basket creation endpoint). I introduced changes to the errors used by v2 which are raised in the `data` module, so the `data` and `errors` modules continue to be versioned, at least for the time being.

jimabramson commented on ecommerce/extensions/payment/helpers.py:
```diff
@@ -3,8 +3,11 @@
+from django.conf import settings
 from django.utils import importlib
 
+from ecommerce.extensions.api.v2 import errors
```
this import feels problematic because
* we're importing from a specific version of the public api into a package that is potentially consumed by other versions
* more generally, we have dependencies in both directions between api/ and supporting packages such as payment/ which will eventually lead to circular import headaches.
in this particular case it seems like it might be easy enough to just move the definition of processor-related errors into the payment package.

clintonb replied:
I agree on all points.

rlucioni replied:
Good points @jimabramson. I've moved processor-related exceptions to the payment app, and I've also moved the 'data' module and API-related exceptions to the root `api` package.

jimabramson commented on ecommerce/extensions/api/v2/tests/test_views.py:
```diff
@@ -0,0 +1,466 @@
+        if checkout:
+            if requires_payment:
+                self.assertIsNone(response.data[AC.KEYS.ORDER])
+                self.assertIsNotNone(response.data[AC.KEYS.PAYMENT_DATA][AC.KEYS.PAYMENT_PROCESSOR_NAME])
```
any reason not to assert that this value matches `payment_processor_name` ?

rlucioni replied:
My reasoning was that it's the API's responsibility to make sure this information is organized and returned correctly, but not to generate or otherwise validate it (that's done by the selected payment processor). Do you think there's value in defining a dummy processor and using it in the context of these tests?

clintonb commented on ecommerce/extensions/analytics/tests/test_analytics.py:
```diff
@@ -4,19 +4,20 @@
 
 
-class AnalyticsTests(OrdersIntegrationTests):
+class AnalyticsTests(CreateBasketViewTests):
```
Unless we need to run the tests in `CreateBasketViewTests` twice, move `assert_successful_basket_creation` to a mixin. Subclass the mixin instead of `CreateBasketViewTests`.

clintonb replied:
The mixin can live outside of the versioned packages.

rlucioni replied:
Fixed, thanks for catching this.

clintonb commented on ecommerce/extensions/api/v2/tests/test_views.py:
```diff
@@ -0,0 +1,466 @@
+    NAME = "dummy-2"
+
+
+class ListPaymentProcessorsViewTests(TestCase, UserMixin):
```
Minor: `PaymentProcessorsListViewTests`

rlucioni replied:
Updated.

clintonb commented on ecommerce/extensions/api/v2/data.py:
```diff
@@ -0,0 +1,93 @@
+OrderTotalCalculator = get_class('checkout.calculators', 'OrderTotalCalculator')
+
+
+def get_basket(user):
```
These first three functions are the exact same as the ones in `v1`. This should be moved to the root `api` package.

rlucioni replied:
Fixed. I was under the impression yesterday that this should be versioned along with the errors, but coming back to it today I realize that's not necessary.

clintonb commented on ecommerce/extensions/api/v2/views.py:
```diff
@@ -0,0 +1,352 @@
+            # returned by this endpoint, simply returning the order number will suffice for now.
+            response_data[AC.KEYS.ORDER] = {AC.KEYS.ORDER_NUMBER: order.number}
+        else:
+            payment_data = {AC.KEYS.PAYMENT_PROCESSOR_NAME: payment_processor.NAME}
```
This can be a single call:

```python
payment_data = {
  AC.KEYS.PAYMENT_PROCESSOR_NAME: ...,
  AC.KEYS.PAYMENT_FORM_DATA: ...,
}
```

rlucioni replied:
Updated.

clintonb commented on ecommerce/extensions/api/v2/views.py:
```diff
@@ -0,0 +1,352 @@
+        )
+
+
+class ListOrderView(ListAPIView):
```
`OrderListView`

rlucioni replied:
Updated.

clintonb commented on ecommerce/extensions/api/v2/views.py:
```diff
@@ -0,0 +1,352 @@
+        return self.request.user.orders.order_by('-date_placed')
+
+
+class RetrieveOrderView(RetrieveAPIView):
```
`OrderRetrieveView`

rlucioni replied:
Updated.

clintonb commented on ecommerce/extensions/api/v2/views.py:
```diff
@@ -0,0 +1,352 @@
+    lookup_field = 'basket_id'
+
+
+class ListPaymentProcessorsView(ListAPIView):
```
`PaymentProcessorListView`

rlucioni replied:
Updated. I've also updated the other view names to match the `{resource}{action}View` convention.

clintonb commented:
Looks great. Most of my comments are naming-related.

jimabramson commented on ecommerce/extensions/api/data.py:
```diff
@@ -38,16 +45,38 @@ def get_product(sku):
+        u"Preparing to place order [%s] for the contents of basket [%d]",
+        number,
+        basket.id,
+    )
```
i think this logging statement should be in BasketCreateView._checkout instead of here.

rlucioni replied:
Updated.

jimabramson commented on ecommerce/extensions/api/v2/views.py:
```diff
@@ -0,0 +1,360 @@
+        else:
+            return self._report_bad_request(
+                api_exceptions.SKU_NOT_FOUND_DEVELOPER_MESSAGE,
+                api_exceptions.SKU_NOT_FOUND_USER_MESSAGE
```
minor: this message seems a little inaccurate for this condition.

rlucioni replied:
Good point. I've replaced this messaging with something more precise.

jimabramson commented on ecommerce/extensions/payment/tests/test_processors.py:
```diff
@@ -188,14 +188,14 @@ def _assert_order_parameters(self, order, receipt_page_url=None, cancel_page_url
                 receipt_url=settings.PAYMENT_PROCESSOR_CONFIG['cybersource']['receipt_page_url'],
                 course_key=self.attribute_value.value,
-                order_number=self.order.number
+                order_number=self.basket.id
```
i'm confused by how this differs from the definition of `_generate_receipt_url` which uses `order.id` (and also that the url parameter is still named after the order number)

rlucioni replied:
Since @clintonb's completely overhauling the payment processors in #58, I would pay little heed to any payment processor changes included in this PR. We agreed that I'd make the minimal set of changes required to allow payment parameters to be generated using baskets instead of orders. I didn't want to get lost in the weeds fixing up something that's going to be ripped out very soon.

jimabramson replied:
totally fine with that.

jimabramson commented:
@renzo done reviewing - just one last concern about _generate_receipt_url, otherwise lgtm!

clintonb commented on ecommerce/extensions/fulfillment/api.py:
```diff
@@ -10,7 +10,7 @@
 from django.utils import importlib
 
-from ecommerce.extensions.fulfillment import errors
+from ecommerce.extensions.fulfillment import exceptions as fulfillment_exceptions
```
What's wrong with simply `exceptions`?

rlucioni replied:
This is a pattern I've followed throughout out of a desire to be very clear in the code about which `exceptions` module was being used. Granted, if only one such module is being imported, it's probably clear enough that this kind of aliasing isn't necessary. I'll revert back to `exceptions`.

clintonb commented on ecommerce/extensions/fulfillment/exceptions.py:
```diff
@@ -1,30 +1,27 @@
-"""Errors thrown by the Fulfillment API."""
+"""Exceptions and error messages used by the fulfillment API."""
```
`s/API/module`

rlucioni replied:
Updated.

jimabramson commented:
:+1:

clintonb commented:
:+1: I have minor naming concerns.

rlucioni commented:
Thanks for the review @clintonb and @jimabramson! I'll merge this once the current build passes.
